### PR TITLE
Capstone Vision for 2021 Season

### PR DESCRIPTION
**Pull request by litehed**
# Pull Requests

Please note that we accept pull requests from anyone, but that does not mean it will be merged.

## What kind of change does this PR introduce?
* Feature
This PR adds a pipeline and detector people can use in order to find the position of their capstones. The pipeline will look for the largest contour and return its center point. The detector will then determine the position of the objects based on the center points relative to their position on the screen.

## Did this PR introduce a breaking change?
_A breaking change includes anything that breaks backwards compatibility either at compile or run time._
* No

__Please make sure your PR satisfies the requirements of [the contributing page](CONTRIBUTING.md)__

### CODE REVIEW

**Comment by serivesmejia** [comment]
I recommend renaming the pipeline from "Capstone" to "TeamShippingElement" ("TSE"?)

**Comment by litehed** [comment]
> I recommend renaming the pipeline from "Capstone" to "TeamShippingElement" ("TSE" for a shorter one?)

I was thinking about that but then I realized FFTSEDetector would not look like the best name for a class. Or we could just separate our vision out into different packages for each season and I could Remove the FF.

**Review by JarnaChao09 — COMMENTED** [review]
Everything else looks fine, the math checks out (assuming raw moments), should be ready to merge once the 3.2 and 1.6 get resolved

**Comment by litehed** [comment]
> Everything else looks fine, the math checks out (assuming raw moments), should be ready to merge once the 3.2 and 1.6 get resolved

Yeah, I was just messing around and forgot to change it back, so now it is less than 1/3 of the screen is left and greater than 2/3 is right. Also, should I make the value based on a percentage of the screen or just pixels?

**Comment by JarnaChao09** [comment]
> > Everything else looks fine, the math checks out (assuming raw moments), should be ready to merge once the 3.2 and 1.6 get resolved
> 
> Yeah, I was just messing around and forgot to change it back, so now it is less than 1/3 of the screen is left and greater than 2/3 is right. Also, should I make the value based on a percentage of the screen or just pixels?

I would say store it in pixels and just make helper functions that take in a percentage instead and then have it calculate it based on that. Some users may want more fine control while others will be like "ehhh screw it, set it to 50%"

**Comment by litehed** [comment]
> 

Just made a change.
Is this what you mean?

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/FFCapstonePipeline.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        Core.inRange(output, low, high, mask);
+        // Puts the mask over the original image
+        Core.bitwise_and(input, input, frame, mask);
+
+        // Blurs the image to smooth it out and reduce unwanted pixels
+        Imgproc.GaussianBlur(mask, mask, new Size(11, 15), 0.0);
+
+        // Initialize the Mat and ArrayList and finds the contours on the image
+        List<MatOfPoint> contours = new ArrayList<>();
+        Mat hierarchy = new Mat();
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/FFCapstonePipeline.java:59]: Does this get released at some point?

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/FFCapstonePipeline.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        // the biggest contour and saves it to the variable biggest.
+        if (hierarchy.size().height > 0 && hierarchy.size().width > 0) {
+            MatOfPoint biggest = new MatOfPoint();
+            for (int index = 0; index >= 0; index = (int) hierarchy.get(0, index)[0]) {
+                Imgproc.drawContours(frame, contours, index, new Scalar(88, 0, 0), 2);
+                if (index == 0)
+                    biggest = contours.get(index);
+                else if (contours.get(index).size().area() > contours.get(index - 1).size().area())
+                    biggest = contours.get(index);
+            }
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/FFCapstonePipeline.java:72]: You should clarify what is meant here by "contours."

**Review by JarnaChao09 — CHANGES_REQUESTED** [review]
Thresholds are always something that should be left as transparent as possible. Giving the user options to define their own thresholds is important. I would look into some way of giving the user a way to define their own threshold calculations, whether it is through a new constructor, or maybe using lambdas for threshold calculation instead.

**Comment by litehed** [comment]
> Thresholds are always something that should be left as transparent as possible. Giving the user options to define their own thresholds is important. I would look into some way of giving the user a way to define their own threshold calculations, whether it is through a new constructor, or maybe using lambdas for threshold calculation instead. 

In line 220 of the detector I have the option to set the threshold. Is that what you mean or do you mean setting the logic of the threshold as well?

**Comment by JarnaChao09** [comment]
> 

no, I was thinking more so you are able to change the formula used. Instead of width / 3, it would be width / 4 and the user could supply this equation through either a lambda or just doing the calculation in their own function and supplying the result.

**Review by JarnaChao09 — APPROVED** [review]

**Pull request merged**